File: solutions/interval/056-merge-intervals.py
"""
# 056. Merge Intervals
**Medium**

Given a problem that demonstrates key concepts in Interval.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of interval concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply interval methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages interval principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses interval techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using interval method
3. Return the computed result

</details>
"""

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        """
        Merges overlapping intervals in the input list.
        
        Args:
            intervals: List of intervals where each interval is [start, end]
        
        Returns:
            List of merged intervals
            
        Time Complexity: O(n log n) due to sorting
        Space Complexity: O(n) or O(1) depending if we consider output space
        """
        # Handle edge cases
        if not intervals:
            return []
        if len(intervals) == 1:
            return intervals
        
        # Sort intervals based on start time
        intervals.sort(key=lambda x: x[0])
        
        # Initialize result with first interval
        merged = [intervals[0]]
        
        # Iterate through remaining intervals
        for current in intervals[1:]:
            # Get the last interval from merged list
            previous = merged[-1]
            
            # If current interval overlaps with previous
            if current[0] <= previous[1]:
                # Update the end time of previous interval
                previous[1] = max(previous[1], current[1])
            else:
                # No overlap, add current interval to result
                merged.append(current)
        
        return merged

def test_solution():
    """
    Test cases for 056. Merge Intervals.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 056. Merge Intervals")

File: solutions/arrays-hashing/alternatives/056-merge-intervals.js
/**
 * 56. Merge Intervals
 * Medium
 *
 * Merge Intervals - JavaScript Implementation This solution merges overlapping intervals in an array of interval pairs. Time Complexity: O(n log n) due to sorting Space Complexity: O(n) to store the result @param {number[][]} intervals - Array of interval pairs [[start1, end1], [start2, end2], ...] @returns {number[][]} - Merged intervals
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving Merge Intervals is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * Merge Intervals - JavaScript Implementation
 * 
 * This solution merges overlapping intervals in an array of interval pairs.
 * Time Complexity: O(n log n) due to sorting
 * Space Complexity: O(n) to store the result
 * 
 * @param {number[][]} intervals - Array of interval pairs [[start1, end1], [start2, end2], ...]
 * @returns {number[][]} - Merged intervals
 */

const mergeIntervals = function(intervals) {
    // Handle edge cases
    if (!intervals || intervals.length === 0) {
        return [];
    }
    
    if (intervals.length === 1) {
        return intervals;
    }

    // Sort intervals based on start time
    intervals.sort((a, b) => a[0] - b[0]);
    
    // Initialize result array with first interval
    const result = [intervals[0]];
    
    // Iterate through remaining intervals
    for (let i = 1; i < intervals.length; i++) {
        // Get the last interval from result and current interval
        const currentInterval = intervals[i];
        const lastMergedInterval = result[result.length - 1];
        
        // Check if current interval overlaps with last merged interval
        if (currentInterval[0] <= lastMergedInterval[1]) {
            // Merge overlapping intervals by updating the end time
            lastMergedInterval[1] = Math.max(lastMergedInterval[1], currentInterval[1]);
        } else {
            // If no overlap, add current interval to result
            result.push(currentInterval);
        }
    }
    
    return result;
};

/**
 * Helper function to test the implementation
 * @param {number[][]} intervals - Test input
 */
function testMergeIntervals(intervals) {
    console.log('Input:', intervals);
    console.log('Output:', mergeIntervals(intervals));
    console.log('---');
}

// Test cases
function runTests() {
    testMergeIntervals([[1,3], [2,6], [8,10], [15,18]]);
    // Expected output: [[1,6], [8,10], [15,18]]
    
    testMergeIntervals([[1,4], [4,5]]);
    // Expected output: [[1,5]]
    
    testMergeIntervals([[1,4], [2,3]]);
    // Expected output: [[1,4]]
    
    testMergeIntervals([]);
    // Expected output: []
    
    testMergeIntervals([[1,5]]);
    // Expected output: [[1,5]]
    
    testMergeIntervals([[1,4], [0,4]]);
    // Expected output: [[0,4]]
}

// Export the function for use in other modules
module.exports = {
    mergeIntervals
};

// Run tests if this file is being run directly
if (require.main === module) {
    runTests();
}